### DISCUSSION TITLE
How to render lines

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3710] [Q&A]
Filament allows using lines as a primitive type, but I wonder whether it is possible to render them the way I would want, namely to have them affected by the amount of incident light, but not the direction of incident light.  Thus, an unlit model is definitely not what I want, but in the lit model, the tangents attribute is required, suggesting that it will not work like I want either.

**Comment by romainguy** [comment]
Why are tangents not what you want? If you don't use normals the concept of incident light is going to be absolute and only related to the light intensity.

**Reply by jwwalker** [reply]
@romainguy I'm confused.  I thought "tangents attribute" was shorthand for quaternions, which encode normal, tangent, and bitangent.

**Reply by romainguy** [reply]
It is yes but the important part is the normal. The tangent and bitangent are only useful for anisotropic lighting or normal mapping. You can use whatever bi/tangent you want when not using those features.

**Reply by jwwalker** [reply]
@romainguy  I still don't get how I can supply the quaternions without at least implicitly also supplying normals.

**Reply by romainguy** [reply]
Sure but see my first reply. The amount of incident light is determined by the light intensity and it's incident angle at the surface, which is why we need the normal. If you don't want to provide normals all you can do is color the lines as a function of the light intensity (which you can do yourself if needed). But the light will have equal color everywhere.

**Reply by jwwalker** [reply]
@romainguy I'm not sure what you mean by "which you can do yourself if needed".  And I don't think it should necessarily be true that "the light will have equal color everywhere", because even without incident angles, one could account for the inverse square law for point and spot lights.

I suppose part of the problem may be that Filament is all about being "physically based", and perfectly thin lines don't exist in the physical world.

**Reply by romainguy** [reply]
With point and spot lights yes distance would be taken into account. But the IBL and directional light would just apply constant light everywhere.

Filament being physically based doesn't matter here.

**Reply by jwwalker** [reply]
@romainguy If you say that distance from point and spot lights would be taken into account, then I take it you are not suggesting that I use the unlit material model, which is supposed to "turn off all lighting computations".  But if I use the lit material model, I need to provide the quaternions, right?  Still confused.

**Reply by romainguy** [reply]
Yes to use the lit model you need to provide the quaternions.

**Reply by jwwalker** [reply]
@romainguy I think what you're telling me is that Filament just can't render lines the way I want.  Which is understandable, you can't be all things to all people.
